### PR TITLE
Add moist baroclinic wave

### PR DESCRIPTION
**Pull request by jiahe23**
For now, it is simplest to create a new folder for moist simulation and copy paste. It makes it harder to view the modifications for moist. Please `vimdiff` to see the differences.

Co-authored-by: Toby Bischoff <bischtob@users.noreply.github.com>
Co-authored-by: Anna Jaruga <trontrytel@users.noreply.github.com>

### CODE REVIEW

**Comment by bischtob** [comment]
@dennisYatunin, can you turn this into a working implicit case?

**Comment by dennisYatunin** [comment]
Can we try to avoid duplicating files, especially things like `staggered_nonhydrostatic_model.jl` and `driver.jl`? I'm not sure why you need to duplicate everything in a `sphere_with_moisture` directory. I think all you need to add moisture is the following:
1. Generalize the pressure functions at the top of `staggered_nonhydrostatic_model.jl`, so that there are both dry and moist versions. I think we want something like `pressure!(ᶜp, Y)`, where there are `if` statements in the function that do different things depending on whether `:ρq in propertynames(Y.c)`, and depending on whether `:ρθ`, `:ρe`, or `:ρe_int` is in `propertynames(Y.c)`. Later, we can define similar functions for the Jacobian, like `∂pressure∂ρ!` and `∂pressure∂w!`.
2. Replace all the calls to the old pressure functions in the hybrid direction with calls to the new pressure function.
3. Add a `moist_baroclinic_wave_rhoe.jl` file to `hybrid/sphere/`. Define a new `center_initial_condition` function in this file that includes `ρq`, and define an `additional_remaining_tendency!` that sets the tendency for `ρq`.

**Comment by bischtob** [comment]
@dennisYatunin, I added the changes. Please pull and make the adjustments. Please fix anything I got wrong.

**Comment by dennisYatunin** [comment]
bors try

**Review by dennisYatunin — APPROVED** [review]

File: examples/hybrid/sphere/baroclinic_wave_utilities.jl [hunk 1]
```diff
@@ -1,122 +1,149 @@
...
+    H = R_d * T_0 / grav
+    z_t = FT(15e3)
+    λ_c = FT(20)
+    ϕ_c = FT(40)
+    d_0 = R / 6
+    V_p = FT(1)
+
+    # Constants required for moist initial conditions
+    p_w = FT(3.4e4)
+    q_t = FT(1e-12) # TODO: why not 0?
```
[the remark this answers is not in the document]
Comment (bischtob) [reply examples/hybrid/sphere/baroclinic_wave_utilities.jl:50]: cause that is the setup from the paper.

[the remark this answers is not in the document]
Comment (bischtob) [reply examples/hybrid/sphere/baroclinic_wave_utilities.jl:50]: they don't use 0

[the remark this answers is not in the document]
Comment (dennisYatunin) [reply examples/hybrid/sphere/baroclinic_wave_utilities.jl:50]: Ok, I'll remove that comment, then.

**Review by dennisYatunin — APPROVED** [review]

**Comment by dennisYatunin** [comment]
bors try

**Comment by dennisYatunin** [comment]
bors try

**Comment by dennisYatunin** [comment]
bors try

**Comment by dennisYatunin** [comment]
bors try

File: examples/hybrid/sphere/baroclinic_wave_utilities.jl [hunk 2]
```diff
@@ -1,122 +1,151 @@
...
+    # T = T_v * (1 + q_tot) / (1 + q_tot * Planet.molmass_ratio(params))
+    # This is the actual formula, which would be consistent with TD.
+
+    # Initial values computed from the thermodynamic state
+    ts = TD.PhaseEquil_pTq(params, p, T, q_tot)
+    ρ = TD.air_density(ts)
     if ᶜ𝔼_name === Val(:ρθ)
-        ρθ = ρ * θ(lat, z)
-        return (; ρ, ρθ, uₕ)
+        ᶜ𝔼_kwarg = (; ρθ = ρ * TD.liquid_ice_pottemp(ts))
```
[the remark this answers is not in the document]
Comment (dennisYatunin) [reply examples/hybrid/sphere/baroclinic_wave_utilities.jl:115]: Yeah, the `PhaseEquil_ρθq` and `PhaseNonEquil_ρθq` constructors expect the liquid-ice potential temperature as input (`PhaseDry_ρθ` doesn't matter in this case). So, we either need to use the liquid-ice potential temperature as our prognostic variable, or we need to convert the dry potential temperature to the liquid-ice potential temperatures every time we need to get the thermodynamic state.

**Pull request merged**